File: chronicle-android/app/src/test/java/com/chronicle/app/brain/CapGraphNodesTest.kt
package com.chronicle.app.brain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class CapGraphNodesTest {

    @Test
    fun capGraphNodes_keepsPinnedAndHighDegree() {
        val nodes = (1..20).map { i ->
            GraphNode(
                id = "n$i",
                kind = "concept",
                label = "N$i",
                pinned = i == 1,
                weight = 1.0,
            )
        }
        val edges = (2..20).map { i ->
            GraphEdge(from = "n1", to = "n$i", rel = "related")
        } + listOf(GraphEdge(from = "n2", to = "n3", rel = "related"))
        val graph = BrainGraph(generated = "t", nodes = nodes, edges = edges)
        val capped = capGraphNodes(graph, limit = 5)
        assertTrue(capped.capped)
        assertEquals(5, capped.nodes.size)
        assertTrue(capped.nodes.any { it.id == "n1" && it.pinned })
        assertEquals(20, capped.totalNodes)
    }

    @Test
    fun capGraphNodes_noopUnderLimit() {
        val nodes = listOf(
            GraphNode(id = "a", kind = "concept", label = "A"),
            GraphNode(id = "b", kind = "concept", label = "B"),
        )
        val graph = BrainGraph(
            generated = "t",
            nodes = nodes,
            edges = listOf(GraphEdge(from = "a", to = "b", rel = "manual")),
        )
        val capped = capGraphNodes(graph, limit = 10)
        assertFalse(capped.capped)
        assertEquals(2, capped.nodes.size)
        assertEquals(1, capped.edges.size)
    }
}

File: chronicle-pc/frontend/src/brain/capGraphNodes.test.ts
import { describe, expect, it } from 'vitest'
import type { BrainGraph } from '../api/types'
import { capGraphNodes } from './capGraphNodes'

function makeGraph(n: number): BrainGraph {
  const nodes = Array.from({ length: n }, (_, i) => ({
    id: `n${i}`,
    kind: i % 2 === 0 ? 'topic' : 'entry',
    label: `Node ${i}`,
    weight: i,
    pinned: i === 0,
    ts: `2026-01-${String((i % 28) + 1).padStart(2, '0')}T12:00:00Z`,
  }))
  const edges = Array.from({ length: Math.max(0, n - 1) }, (_, i) => ({
    from: `n${i}`,
    to: `n${i + 1}`,
    rel: 'related',
  }))
  // Hub: connect n0 to many for degree score
  for (let i = 2; i < Math.min(n, 20); i++) {
    edges.push({ from: 'n0', to: `n${i}`, rel: 'related' })
  }
  return { version: 1, nodes, edges }
}

describe('capGraphNodes', () => {
  it('returns all nodes when under the cap', () => {
    const graph = makeGraph(10)
    const capped = capGraphNodes(graph, 500)
    expect(capped.capped).toBe(false)
    expect(capped.nodes).toHaveLength(10)
    expect(capped.totalNodes).toBe(10)
  })

  it('caps to limit and keeps pinned nodes', () => {
    const graph = makeGraph(50)
    const capped = capGraphNodes(graph, 15)
    expect(capped.capped).toBe(true)
    expect(capped.nodes).toHaveLength(15)
    expect(capped.totalNodes).toBe(50)
    expect(capped.nodes.some((n) => n.id === 'n0' && n.pinned)).toBe(true)
    const ids = new Set(capped.nodes.map((n) => n.id))
    for (const e of capped.edges) {
      expect(ids.has(e.from)).toBe(true)
      expect(ids.has(e.to)).toBe(true)
    }
  })

  it('filters hidden nodes before capping', () => {
    const graph = makeGraph(5)
    graph.nodes[1]!.hidden = true
    const capped = capGraphNodes(graph, 500)
    expect(capped.totalNodes).toBe(4)
    expect(capped.nodes.find((n) => n.id === 'n1')).toBeUndefined()
  })
})
